core/lib/zksync_core/src/api_server/tx_sender/proxy.rs

```diff
@@ -1,22 +1,110 @@
+use tokio::sync::{watch, RwLock};
+use zksync_dal::ConnectionPool;
 use zksync_types::{
     api::{BlockId, Transaction, TransactionDetails, TransactionId},
     l2::L2Tx,
-    H256,
+    Address, Nonce, H256,
 };
 use zksync_web3_decl::{
     jsonrpsee::http_client::{HttpClient, HttpClientBuilder},
     namespaces::{EthNamespaceClient, ZksNamespaceClient},
     RpcResult,
 };
 
+#[derive(Debug, Clone, Default)]
+pub(crate) struct TxCache {
+    inner: Arc<RwLock<TxCacheInner>>,
+}
+
+#[derive(Debug, Default)]
+struct TxCacheInner {
+    tx_cache: HashMap<H256, L2Tx>,
+    nonces_by_account: HashMap<Address, BTreeSet<Nonce>>,
+}
+
+impl TxCache {
+    async fn push(&self, tx: L2Tx) {
+        let mut inner = self.inner.write().await;
+        inner
+            .nonces_by_account
+            .entry(tx.initiator_account())
+            .or_default()
+            .insert(tx.nonce());
+        inner.tx_cache.insert(tx.hash(), tx);
+    }
+
+    async fn get_tx(&self, tx_hash: H256) -> Option<L2Tx> {
+        self.inner.read().await.tx_cache.get(&tx_hash).cloned()
+    }
+
+    async fn get_nonces_for_account(&self, account_address: Address) -> Vec<Nonce> {
+        let inner = self.inner.read().await;
+        if let Some(nonces) = inner.nonces_by_account.get(&account_address) {
+            nonces.iter().copied().collect()
+        } else {
+            vec![]
+        }
+    }
+
+    async fn remove_tx(&self, tx_hash: H256) {
+        self.inner.write().await.tx_cache.remove(&tx_hash);
+        // We intentionally don't change `nonces_by_account`; they should only be changed in response to new miniblocks
+    }
+
+    async fn run_updates(
+        self,
+        pool: ConnectionPool,
+        stop_receiver: watch::Receiver<bool>,
+    ) -> anyhow::Result<()> {
+        const UPDATE_INTERVAL: Duration = Duration::from_secs(1);
+
+        loop {
+            if *stop_receiver.borrow() {
+                return Ok(());
+            }
+
+            let addresses: Vec<_> = {
+                // Split into 2 statements for readability.
+                let inner = self.inner.read().await;
+                inner.nonces_by_account.keys().copied().collect()
+            };
+            let mut storage = pool.access_storage_tagged("api").await?;
+            let nonces_for_accounts = storage
+                .storage_web3_dal()
+                .get_nonces_for_addresses(&addresses)
+                .await?;
+            drop(storage); // Don't hold both `storage` and lock on `inner` at the same time.
+
+            let mut inner = self.inner.write().await;
+            inner.nonces_by_account.retain(|address, account_nonces| {
+                let stored_nonce = nonces_for_accounts
+                    .get(address)
+                    .copied()
+                    .unwrap_or(Nonce(0));
+                // Retain only nonces exceeding the stored one.
+                *account_nonces = account_nonces.split_off(&(stored_nonce + 1));
```

Comment:
I think there's an off-by-one error (introduced by me, sorry). `stored_nonce` is actually the next expected nonce for the account (e.g., if the corresponding storage slot is empty – i.e., we have `Nonce(0)` – it means that the *next* nonce should be 0). So here, it should be `split_off(stored_nonce)`. Ideally, this should be covered by a unit test 🙃